src/ipc/ipc4/dai.c

```diff
@@ -30,6 +30,49 @@
 
 LOG_MODULE_DECLARE(ipc, CONFIG_SOF_LOG_LEVEL);
 
+int dai_set_link_hda_config(uint16_t *link_config,
+							uint32_t dai_type,
+							uint32_t dir,
+							struct ipc_config_dai *common_config,
+							const void *spec_config)
+{
+#if defined(CONFIG_ACE_VERSION_2_0)
+	struct ipc4_copier_module_cfg *copier_config =
+	    (struct ipc4_copier_module_cfg *) common_config->host_dma_config;
+	struct ipc4_audio_format *out_fmt = (struct ipc4_audio_format *) common_config->out_fmt;
+	struct sof_ipc_dai_config *sof_cfg;
+	union hdalink_cfg link_cfg;
+
+	switch (dai_type) {
+	case SOF_DAI_INTEL_SSP:
+		sof_cfg = (struct sof_ipc_dai_config *)spec_config;
+		link_cfg.full = 0;
+		link_cfg.part.dir = dir;
+		link_cfg.part.stream = common_config->host_dma_config->stream_id;
+		break;
+	case SOF_DAI_INTEL_DMIC:
+		link_cfg.full = 0;
+		if (out_fmt->depth == IPC4_DEPTH_16BIT) {
+			/* 16bit dmic packs two 16bit samples into single 32bit word
+			 * fw needs to adjust channel count to match final sample
+			 * group size
+			 */
+			link_cfg.part.hchan = (out_fmt->channels_count - 1) / 2;
+		} else {
+			link_cfg.part.hchan = out_fmt->channels_count - 1;
+		}
+		link_cfg.part.stream = common_config->host_dma_config->stream_id;
+		break;
+	default:
+		/* other types of DAIs not handled for now */
```

Comment:
Best to log some failure here since config is coming from host SW.

Reply:
Log added